zerver/tests/test_tusd.py

```diff
@@ -0,0 +1,185 @@
+import os
+
+from django.conf import settings
+
+from zerver.lib.test_classes import ZulipTestCase
+from zerver.lib.test_helpers import (
+    create_s3_buckets,
+    use_s3_backend,
+)
+from zerver.lib.upload.s3 import upload_image_to_s3
+
+
+class ZephyrTest(ZulipTestCase):
+    def test_tusd_auth(self) -> None:
+        body = {
+            "Upload": {
+                "ID": "",
+            },
+        }
+        result = self.client_post(
+            f"/tusd/hooks?secret={settings.SHARED_SECRET}",
+            body,
+            content_type="application/json",
+            HTTP_HOOK_NAME="pre-create",
+        )
+        self.assert_json_error_contains(
+            result, "Not logged in: API authentication or user session required", 401
+        )
+
+    def test_tusd_pre_create_hook(self) -> None:
+        self.login("hamlet")
+        body = {
+            "Upload": {
+                "ID": "",
+                "IsFinal": False,
+                "IsPartial": False,
+                "MetaData": {
+                    "filename": "jupyterlab-galata-report.zip",
+                    "filetype": "application/zip",
+                    "name": "jupyterlab-galata-report.zip",
+                    "type": "application/zip",
+                },
+                "Offset": 0,
+                "PartialUploads": None,
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 - 100,
+                "SizeIsDeferred": False,
+            },
+        }
+        result = self.client_post(
+            f"/tusd/hooks?secret={settings.SHARED_SECRET}",
+            body,
+            content_type="application/json",
+            HTTP_HOOK_NAME="pre-create",
+        )
+        self.assert_json_success(result)
+
+    def test_file_too_big_failure(self) -> None:
+        self.login("hamlet")
+        body = {
+            "Upload": {
+                "Size": settings.MAX_FILE_UPLOAD_SIZE * 1024 * 1024 + 100,
+            },
+        }
+        with self.settings(MAX_FILE_UPLOAD_SIZE=0):
+            result = self.client_post(
+                f"/tusd/hooks?secret={settings.SHARED_SECRET}",
+                body,
+                content_type="application/json",
+                HTTP_HOOK_NAME="pre-create",
+            )
```

Comment:
It seems redundant to make the upload larger than the limit, and then test with the limit set to 0 -- you only need to do one of those things.  The former makes more sense to me as a test.

Reply:
updated.